server/src/services/QwixxLogic.ts

```diff
@@ -120,25 +184,60 @@ export default class QwixxLogic {
       player.submissionCount === 1 &&
       !this._dice.validColouredNumbers[colourToMark]?.includes(num)
     ) {
-      throw new Error(
-        "Number selected doesn't equal to sum of white die and coloured die."
-      );
-    }
-
-    if (!player.markNumber(colourToMark, num)) {
-      throw new Error("Invalid move.");
+      return {
+        isValid: false,
+        errorMessage: new Error(
+          "Number selected doesn't equal to sum of white die and coloured die."
+        ),
+      };
     }
 
-    if (
-      (player === this.activePlayer && player.submissionCount === 2) ||
-      (player !== this.activePlayer && player.submissionCount === 1)
-    ) {
-      player.markSubmitted();
+    /*add check for number being lower or higher than last checked number */
+    if (colourToMark === "red" || colourToMark === "yellow") {
+      if (num <= highestMarkedNumber) {
+        return {
+          isValid: false,
+          errorMessage: new Error(
+            "Number must be above the last marked number"
+          ),
+        };
+      }
+    } else if (colourToMark === "green" || colourToMark === "blue") {
+      if (num >= lowestMarkedNumber) {
+        return {
+          isValid: false,
+          errorMessage: new Error(
+            "Number must be below the last marked number"
+          ),
+        };
+      }
     }
 
-    this.processPlayersSubmission();
+    return {
+      isValid: true,
+      errorMessage: null,
+    };
+  }
 
-    return this.serialize();
+  public validMoveAvailable(): Record<string, boolean> {
+    const playerMoveAvailable: Record<string, boolean> = {};
+
+    for (const player of this._playersArray) {
+      let hasValidMove = false;
+
+      for (let row of ["red", "yellow", "green", "blue"]) {
+        for (let num = 2; num <= 12; num++) {
+          const validationResult = this.validMove(player.name, row, num);
```

Comment:
Do all the conditions inside validMove apply for validMoveAvailable? The validMoveAvailable() is used for checking if a player has a valid move after rolling a dice, right? So maybe conditions like "if dice rolled" or "hasSubmittedChoice" wouldn't apply. So maybe we shouldn't use validMove() like that, or we should break down the parts of validMove() to make them more reusable.